### PR TITLE
Jordy add roles controller unit tests

### PR DESCRIPTION
**Pull request by j-corp-25**
# Description

Unit tests for rolesController

## Related PRS (if any):
N/A

## Main changes explained:

- unit tests for `createNewRole`
- unit tests for `deleteRoleById`
- unit tests for `getAllRoles`
- unit tests for `getRoleById`
- unit tests for `updateRoleById`

## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. run `npm run test roles -- --verbose`

## Screenshots

PRTest

## Note:

Check that all tests are passing
N/A

### CODE REVIEW

**Review by Kavil-Jain-514 — APPROVED** [review]
All test cases have passed

image

**Review by jinxyou — APPROVED** [review]
Screenshot 2024-06-08 153336
Hi I have done the testing according to the instruction. It seems no problem. Great work!

**Comment by akshit0211** [comment]
Hi Jordy, all test cases passed. Great work!

all test cases passed

**Review by akshit0211 — APPROVED** [review]

**Review by Logeshwari-Renu — APPROVED** [review]
All the test cases are passed.! 

PR981

**Review by Wenboooooo — APPROVED** [review]
All tests have passed. Looks good to me.

**Review by deepthikannan — APPROVED** [review]
image
All test cases passed.

**Review by KaushikMreddy — APPROVED** [review]
Hi Jordy, all your test cases are working, please try to add a case where deleting a role fails if appropriate for the function.

image

File: src/controllers/rolesController.js [hunk 1]
```diff
@@ -1,24 +1,23 @@
-const UserProfile = require("../models/userProfile");
-const cache = require("../utilities/nodeCache")();
-const { hasPermission } = require("../utilities/permissions");
+const UserProfile = require('../models/userProfile');
```
Comment (TXMO-dev) [inline src/controllers/rolesController.js:1]: Great job on the recent code! It shows a solid understanding of Node.js and Express for handling roles and permissions. Here are a few suggestions that could further enhance the readability, maintainability, and performance of your code:
1. Improve Error Handling:

Issue: Error handling could be more standardized and descriptive.
Suggestion: Consider using a centralized error handling middleware to maintain consistency across your application and provide more detailed error messages.

```
const errorHandler = (err, req, res, next) => {
  res.status(err.status || 500).send({ error: err.message || 'Internal Server Error' });
};
```
2. Consistent Asynchronous Code:

Issue: Mixing callback-based and async/await-based code can lead to inconsistencies.
Suggestion: Refactor your code to consistently use async/await.

```
const getAllRoles = async (req, res, next) => {
  try {
    const results = await Role.find({});
    res.status(200).send(results);
  } catch (error) {
    next(error);
  }
};

const createNewRole = async (req, res, next) => {
  try {
    const newRole = new Role({
      roleName: req.body.roleName,
      permissions: req.body.permissions,
    });
    const savedRole = await newRole.save();
    res.status(201).send(savedRole);
  } catch (error) {
    next(error);
  }
};
```
Enhanced Input Validation:

Issue: Current input validation is minimal and could be more robust.
Suggestion: Utilize a validation library like Joi or express-validator for more comprehensive input validation.

Summary of Changes:
Centralized Error Handling: Adds middleware for consistent error handling.
Consistent Async/Await Usage: Refactors functions to use async/await.
Permission Check Middleware: Moves permission checks to middleware.
Enhanced Input Validation: Adds middleware for more robust input validation.
These improvements will make your codebase more modular, readable, and easier to maintain. Keep up the fantastic work!

**Review by TXMO-dev — CHANGES_REQUESTED** [review]

**Review by aaryaneil — APPROVED** [review]
Tested the PR and all the test cases passed.
Screenshot 2024-06-15 at 4 57 39 AM

**Review by shradac — APPROVED** [review]
All tests passed. The tests cover all specified scenarios. Good Work!

**Comment by mianmiantea2019** [comment]
All tests passed.  Good work!
image

**Comment by sahithidhulipala9** [comment]
All test cases passed! 
PR 981

**Review by sahithidhulipala9 — APPROVED** [review]

**Review by Rishitha1309 — APPROVED** [review]
All test cases passed successfully
PR981

**Review by MinSun0814 — APPROVED** [review]
image

**Review by nikhilpittala16 — APPROVED** [review]
I have tested the PR and all the test cases have passed.
Screenshot (53)

**Review by gowthamdongari — APPROVED** [review]
All test cases working as expected 
image

**Pull request merged**